File: pkg/orchestrator/machine.go
/*
* This file is part of Celestial (https://github.com/OpenFogStack/celestial).
* Copyright (c) 2021 Tobias Pfandzelter, The OpenFogStack Team.
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, version 3.
*
* This program is distributed in the hope that it will be useful, but
* WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
* General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program. If not, see <http://www.gnu.org/licenses/>.
**/

package orchestrator

import (
	"context"
	"io"
	"math/rand"
	"net"
	"sync"
	"sync/atomic"
	"time"

	"github.com/pkg/errors"

	"github.com/firecracker-microvm/firecracker-go-sdk"
	log "github.com/sirupsen/logrus"

	"github.com/OpenFogStack/celestial/pkg/commons"
)

type remoteMachine struct {
	host *host
}

type localmachine struct {
	active bool

	links map[string]*link

	name string

	// and some firecracker reference
	m               *firecracker.Machine
	drivePath       string
	kernelImagePath string
	htEnabled       bool
	memSizeMiB      uint64
	vCPUCount       uint64
	bandwidth       uint64
	bootparams      string
	outFile         io.Writer
	errFile         io.Writer
	initialized     bool

	// and some networking stuff
	tapName        string
	chainName      string
	ipBlockSet     string
	ip             string
	gateway        string
	mac            string
	hostInterface  string
	nextHandle     int64
	netInitialized bool

	sync.RWMutex
}

type machine struct {
	isLocal bool

	*remoteMachine
	*localmachine

	id        id
	address   net.IP
	network   *net.IPNet
	bandwidth uint64
	name      string
}

// GetGSTID returns the ID of a ground station machine.
func (o *Orchestrator) GetGSTID(name string) (uint64, error) {

	m, err := o.getGSTMachineByName(name)

	if err != nil {
		return 0, errors.WithStack(err)
	}

	return uint64(m.id), nil
}

func (o *Orchestrator) getMachine(m commons.MachineID) (*machine, error) {
	s, ok := o.shells[shellid(m.Shell)]

	if !ok {
		return nil, errors.Errorf("unknown shell: %d", m.Shell)
	}

	s.RLock()
	machine, ok := s.machines[id(m.ID)]
	s.RUnlock()

	if !ok {
		return nil, errors.Errorf("unknown machine: %d in shell %d", m.ID, m.Shell)
	}

	return machine, nil
}

func (o *Orchestrator) getGSTMachineByName(name string) (*machine, error) {
	o.gstLock.RLock()
	machine, ok := o.groundstations[name]
	o.gstLock.RUnlock()

	if !ok {
		return nil, errors.Errorf("unknown groundstation: %s", name)
	}

	return machine, nil
}

// setInactive pauses a firecracker microVM.
func (o *Orchestrator) setInactive(m *localmachine) error {

	m.active = false

	if !m.initialized {
		return nil
	}

	c, cancel := context.WithCancel(context.Background())
	defer cancel()

	err := m.m.PauseVM(c)

	return errors.WithStack(err)
}

// setActive resumes a firecracker microVM.
func (o *Orchestrator) setActive(m *localmachine) error {

	m.active = true

	if !m.initialized {
		err := m.initialize()

		return errors.WithStack(err)
	}

	c, cancel := context.WithCancel(context.Background())
	defer cancel()

	err := m.m.ResumeVM(c)

	return errors.WithStack(err)
}

// CreateMachine creates a new firecracker microVM.
func (o *Orchestrator) CreateMachine(m commons.MachineID, vCPUCount uint64, memSizeMiB uint64, htEnabled bool, bootparams string, kernel string, rootfs string, bandwidth uint64, active bool) error {

	atomic.AddInt64(&o.outstanding, 1)
	defer atomic.AddInt64(&o.outstanding, -1)

	// check if that shell even exists
	s, ok := o.shells[shellid(m.Shell)]

	if !ok {
		return errors.Errorf("unknown shell: %d", m.Shell)
	}

	// check if the machine exists already
	s.RLock()
	if _, ok := s.machines[id(m.ID)]; ok {
		s.RUnlock()
		return errors.Errorf("machine exists already: %d in in shell %d", m.ID, m.Shell)
	}
	s.RUnlock()

	machine := &machine{
		isLocal: true,
		localmachine: &localmachine{
			active:         active,
			links:          make(map[string]*link),
			m:              nil,
			initialized:    false,
			netInitialized: false,
			tapName:        "",
			nextHandle:     0,
			RWMutex:        sync.RWMutex{},
		},
		bandwidth: bandwidth,
		name:      m.Name,
	}

	machine.Lock()
	defer machine.Unlock()

	// create the machine
	err := machine.create(m.ID, m.Shell, m.Name, vCPUCount, memSizeMiB, htEnabled, bandwidth, kernel, rootfs, bootparams, o.networkInterface)

	if err != nil {
		return errors.WithStack(err)
	}

	if m.Shell == -1 {
		o.gstLock.Lock()
		o.groundstations[m.Name] = machine
		o.gstLock.Unlock()
	}

	s.Lock()
	s.machines[id(m.ID)] = machine
	s.Unlock()

	err = o.registerLocal(m.ID, m.Shell, m.Name)

	if err != nil {
		return errors.WithStack(err)
	}

	if active || o.eager {

		// proposal: don't overload the machine at the beginning
		time.Sleep(time.Duration(rand.Intn(o.initDelay)) * time.Second)

		log.Debugf("initializing machine: %d %d %s %v", m.ID, m.Shell, m.Name, active)

		err := machine.localmachine.initialize()

		if err != nil {
			log.Errorf("%+v\n", err)
			// return errors.WithStack(err)
		}

		if !active {
			return o.setInactive(machine.localmachine)
		}
	}

	atomic.AddUint64(&o.created, 1)
	log.Debugf("%d machines created, %d machines outstanding", o.created, o.outstanding)

	return nil
}

// ModifyMachine pauses or resumes a firecracker microVM.
func (o *Orchestrator) ModifyMachine(m commons.MachineID, active bool) error {

	log.Debugf("modifying machine: %d %d %s %v", m.ID, m.Shell, m.Name, active)

	machine, err := o.getMachine(m)

	if err != nil {
		return errors.WithStack(err)
	}

	if !machine.isLocal {
		return errors.Errorf("cannot modify machine %v, is not local", m)
	}

	machine.localmachine.Lock()
	defer machine.localmachine.Unlock()

	if machine.localmachine.active && !active {
		log.Infof("Setting machine %d in shell %d to: inactive", m.ID, m.Shell)
		return o.setInactive(machine.localmachine)
	}

	if !machine.localmachine.active && active {
		log.Infof("Setting machine %d in shell %d to: active", m.ID, m.Shell)
		return o.setActive(machine.localmachine)
	}

	return nil
}

File: celestial/machine.py
#
# This file is part of Celestial (https://github.com/OpenFogStack/celestial).
# Copyright (c) 2021 Tobias Pfandzelter, The OpenFogStack Team.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
#

import typing
import threading as td

from .machine_connector import MachineConnector

DEBUG = False

# update threshold for updating latency in ms
LATENCY_UPDATE_THRESHOLD = 0.5

# update threshold for updating bandwidth in Kbit
BANDWIDTH_UPDATE_THRESHOLD = 1


class Machine:
    def __init__(
        self,
        shell_no: int,
        plane_no: int,
        id: int,
        active: bool,
        machine_connector: MachineConnector,
    ):
        self.shell_no = shell_no
        self.plane_no = plane_no
        self.id = id

        self.active = active
        self.mutex = td.Semaphore(1)

        self.connector = machine_connector

        self.links: typing.Dict[
            Machine, typing.Dict[str, typing.Union[int, float]]
        ] = {}
        self.new_links: typing.Dict[
            Machine, typing.Dict[str, typing.Union[int, float]]
        ] = {}
        self.new_unlinks: typing.Set[Machine] = set()
        self.link_mutex = td.Semaphore(1)

    def reset_links(self) -> None:
        self.link_mutex.acquire()

        self.new_links = {}
        self.new_unlinks = set()

        self.link_mutex.release()

    def set_links(self) -> None:
        self.link_mutex.acquire()

        remove_set: typing.List[typing.Dict[str, int]] = []
        modify_set: typing.List[typing.Dict[str, typing.Union[int, float]]] = []

        for target in self.new_unlinks:
            remove_set.append(
                {
                    "shell": target.shell_no,
                    "sat": target.id,
                }
            )

            if DEBUG:
                print(
                    "removing delay from %s %s to %s %s"
                    % (self.shell_no, self.id, target.shell_no, target.id)
                )

        for target in self.links:
            if target not in self.new_links:
                # remove link
                remove_set.append(
                    {
                        "shell": target.shell_no,
                        "sat": target.id,
                    }
                )

                if DEBUG:
                    print(
                        "removing delay from %s %s to %s %s"
                        % (self.shell_no, self.id, target.shell_no, target.id)
                    )

        for target in self.new_links:
            if target in self.links:
                # modify link
                latency_diff = abs(
                    self.new_links[target]["latency"] - self.links[target]["latency"]
                )
                bandwidth_diff = abs(
                    self.new_links[target]["bandwidth"]
                    - self.links[target]["bandwidth"]
                )

                if (
                    not latency_diff > LATENCY_UPDATE_THRESHOLD
                    and not bandwidth_diff > BANDWIDTH_UPDATE_THRESHOLD
                ):
                    continue

            # add link
            modify_set.append(
                {
                    "shell": target.shell_no,
                    "sat": target.id,
                    "latency": self.new_links[target]["latency"],
                    "bandwidth": self.new_links[target]["bandwidth"],
                }
            )
            if DEBUG:
                print(
                    "modifying delay from %s %s to %s %s to: %d %d"
                    % (
                        self.shell_no,
                        self.id,
                        target.shell_no,
                        target.id,
                        self.new_links[target]["latency"],
                        self.new_links[target]["bandwidth"],
                    )
                )

        try:
            self.connector.modify_links(remove_set=remove_set, modify_set=modify_set)

        except Exception as e:
            print(
                "caught exception while trying to update links for machine %d shell %d:"
                % (self.id, self.shell_no),
                e,
            )

        finally:
            self.links = {}
            for target in self.new_links:
                self.links[target] = self.new_links[target]

        self.link_mutex.release()

    def link(self, target: "Machine", latency: float, bandwidth: int) -> None:
        self.link_mutex.acquire()

        if target in self.new_links:
            if latency < self.new_links[target]["latency"]:
                self.new_links[target] = {"latency": latency, "bandwidth": bandwidth}

        else:
            self.new_links[target] = {"latency": latency, "bandwidth": bandwidth}

        self.link_mutex.release()

    def unlink(self, targets: typing.Set["Machine"]) -> None:
        self.link_mutex.acquire()

        self.new_unlinks = self.new_unlinks.union(targets)

        self.link_mutex.release()

    def set_active(self, block: bool = False) -> None:
        t = td.Thread(target=self.__set_active)
        t.start()

        if block:
            t.join()

    def __set_active(self) -> None:
        self.mutex.acquire()

        if not self.active:
            try:
                self.active = True
                self.connector.modify_machine(self.active)
            except Exception as e:
                self.active = False
                print(
                    "caught exception while trying to set machine active %d shell %d:"
                    % (self.id, self.shell_no),
                    e,
                )

        self.mutex.release()

    def set_inactive(self, block: bool = False) -> None:
        t = td.Thread(target=self.__set_inactive)

        t.start()

        if block:
            t.join()

    def __set_inactive(self) -> None:
        self.mutex.acquire()

        if self.active:
            try:
                self.active = False
                self.connector.modify_machine(self.active)
            except Exception as e:
                self.active = True
                print(
                    "caught exception while trying to set machine inactive %d shell %d:"
                    % (self.id, self.shell_no),
                    e,
                )

        self.mutex.release()
